velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp

```diff
@@ -0,0 +1,381 @@
+    }
+    bytesRead_ += length;
+  }
+
+  std::shared_ptr<gcs::Client> client_;
+  std::string bucket_;
+  std::string key_;
+  std::atomic<int64_t> length_ = -1;
+};
+
+} // namespace
+
+namespace filesystems {
+
+class GCSConfig {
+ private:
+  const Config* FOLLY_NONNULL config_;
+  std::shared_ptr<gc::Credentials> credentials_;
+
+ public:
+  GCSConfig(const Config* config) : config_(config) {
+    std::string cred = config_->get("hive.gcs.credentials", std::string(""));
+    if (!cred.empty()) {
+      credentials_ = gc::MakeServiceAccountCredentials(cred);
+    } else {
+      LOG(WARNING) << "Config hive.gcs.credentials is empty";
+    }
+  }
+
+  std::string endpointOverride() const {
+    return config_->get("hive.gcs.endpoint", std::string(""));
+  }
+
+  std::string scheme() const {
+    return config_->get("hive.gcs.scheme", std::string("https"));
+  }
+
+  std::shared_ptr<gc::Credentials> credentials() const {
+    return credentials_;
+  }
+};
+
+class GCSFileSystem::Impl {
+ public:
+  Impl(const Config* config) : gcsConfig_(config) {}
+
+  ~Impl() = default;
+
+  // Use the input Config parameters and initialize the GCSClient.
+  void initializeClient() {
+    auto options = gc::Options{};
+    std::string scheme = gcsConfig_.scheme();
+    // if (scheme.empty()) scheme = "https";
+    if (scheme == "https") {
+      options.set<gc::UnifiedCredentialsOption>(
+          gc::MakeGoogleDefaultCredentials());
+    } else {
+      options.set<gc::UnifiedCredentialsOption>(gc::MakeInsecureCredentials());
+    }
+    options.set<gcs::UploadBufferSizeOption>(kUploadBufferSize);
+    if (!gcsConfig_.endpointOverride().empty()) {
+      options.set<gcs::RestEndpointOption>(
+          scheme + "://" + gcsConfig_.endpointOverride());
+    }
+    if (gcsConfig_.credentials()) {
+      options.set<gc::UnifiedCredentialsOption>(gcsConfig_.credentials());
+    }
+    client_ = std::make_shared<gcs::Client>(options);
+  }
+
+  std::shared_ptr<gcs::Client> getClient() const {
+    return client_;
+  }
+
+ private:
+  const GCSConfig gcsConfig_;
+  std::shared_ptr<gcs::Client> client_;
+};
+
+GCSFileSystem::GCSFileSystem(std::shared_ptr<const Config> config)
+    : FileSystem(config) {
+  impl_ = std::make_shared<Impl>(config.get());
+}
+
+void GCSFileSystem::initializeClient() {
+  impl_->initializeClient();
+}
+
+std::unique_ptr<ReadFile> GCSFileSystem::openFileForRead(
+    std::string_view path,
+    const FileOptions& /*unused*/) {
+  const std::string gcspath = gcsPath(path);
+  auto gcsfile = std::make_unique<GCSReadFile>(gcspath, impl_->getClient());
+  gcsfile->initialize();
+  return gcsfile;
+}
+
+std::unique_ptr<WriteFile> GCSFileSystem::openFileForWrite(
+    std::string_view path,
+    const FileOptions& /*unused*/) {
+  VELOX_NYI();
+}
+
+void GCSFileSystem::remove(std::string_view path) {
+  if (!isGCSFile(path)) {
+    VELOX_FAIL("File {} is not a valid gcs file", path);
+  }
+
+  // assumption it's a proper path
```

Comment:
nit: `// We assume 'path' is well-formed here.`

Reply:
Done.